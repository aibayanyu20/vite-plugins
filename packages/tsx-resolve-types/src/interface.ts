import type { ParseResult } from '@babel/parser'
import type { File } from '@babel/types'

export type AST = ParseResult<File>

export interface UserOptions {
  /**
   * @description 是否开启emits的类型推断
   * @default true
   */
  emits?: boolean
  /**
   * @description 是否开启props的类型推断
   * @default true
   */
  props?: boolean
}
