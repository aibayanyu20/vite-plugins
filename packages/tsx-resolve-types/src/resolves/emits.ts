export function resolveEmits(emits: Record<string, any>) {
  // TODO
}
