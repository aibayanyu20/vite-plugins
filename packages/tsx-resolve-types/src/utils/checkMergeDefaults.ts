import type { CreateContextType } from './context'
import { traverse } from './traverse'
import { createAst } from './ast'

export function checkMergeDefaults(ctx: CreateContextType) {
  if (!ctx.importMergeDefaults)
    return

  /**
   * 判断是否存在导入了
   */
  let hasImportVue = false
  traverse(ctx.ast, {
    ImportDeclaration(path) {
      if (path.node.source.value === 'vue') {
        const node = path.node.specifiers.find((v) => {
          return v.local.name === '_mergeDefaults'
        })
        if (node)
          hasImportVue = true
      }
    },
  })
  if (!hasImportVue) {
    const myAst = createAst(`import { mergeDefaults as  _mergeDefaults } from 'vue'`, false)
    ctx.ast.program.body.unshift(
      ...myAst.program.body,
    )
  }
}
