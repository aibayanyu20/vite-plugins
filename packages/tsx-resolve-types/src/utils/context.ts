import fs from 'node:fs'
import type { SimpleTypeResolveContext } from '@vue/compiler-sfc'
import type { AST } from '../interface'
import { createAst } from './ast'
import type { GraphContext } from './graphContext'

export interface CreateContextType {
  ctx: SimpleTypeResolveContext
  ast: AST
  source: string
  filepath: string
  importMergeDefaults?: boolean
}

export function createContext(code: string, id: string, graphCtx?: GraphContext): CreateContextType {
  const ast = createAst(code)
  const helper = new Set<string>()
  return {
    ast,
    filepath: id,
    source: code,
    importMergeDefaults: false,
    ctx: {
      filename: id,
      source: code,
      ast: ast.program.body,
      error(node, msg) {
        console.error(msg, node)
        throw new Error(node)
      },
      helper(key) {
        helper.add(key)
        return `_${key}`
      },
      getString(node) {
        return code.slice(node.start!, node.end!)
      },
      propsTypeDecl: undefined,
      propsRuntimeDefaults: undefined,
      propsDestructuredBindings: Object.create(null),
      emitsTypeDecl: undefined,
      isCE: false,
      options: {
        fs: {
          fileExists(_file: string): boolean {
            // 检查文件是否存在
            try {
              const stat = fs.statSync(_file)
              if (stat.isFile()) {
                if (!(/node_modules/.test(_file))) {
                  graphCtx?.add(_file, id)
                  graphCtx?.addGraph(id, _file)
                }

                return true
              }
              return false
            }
            catch (e) {
              return false
            }
          },
          readFile(_file: string): string | undefined {
            return fs.readFileSync(_file, 'utf-8')
          },
        },
      },
    },
  }
}
