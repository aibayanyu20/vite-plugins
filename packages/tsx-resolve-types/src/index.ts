import type { PluginOption } from 'vite'
import { GraphContext } from './utils/graphContext'
import { invalidateTypeCacheId } from './utils/invalidate'
import { getDepModules } from './utils/depModules'

export function tsxResolveTypes(): PluginOption {
  const graphCtx = new GraphContext()
  return {
    name: 'tsx-resolve-types',

    /**
     * 热更新
     */
    handleHotUpdate({ file, server, modules }) {
      if (graphCtx.has(file)) {
        // 清理缓存
        invalidateTypeCacheId(file)
        // 如果存在那么响应的所有的moduleGraph都需要更新
        const ids = graphCtx.get(file)
        const affectedModules = []
        if (ids) {
          for (const id of ids) {
            const mods = getDepModules(id, graphCtx, server.moduleGraph)
            graphCtx.removeCache(id)
            if (mods.length)
              affectedModules.push(...mods)
          }
          return [...modules, ...affectedModules]
        }
      }
      return modules
    },
  }
}
