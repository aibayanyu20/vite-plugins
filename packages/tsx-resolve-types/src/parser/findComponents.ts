import type { CallExpression } from '@babel/types'
import { isCallExpression } from '@babel/types'
import type { AST } from '../interface'
import { traverse } from '../utils/traverse'

function checkIsDefineComponent(node: CallExpression) {
  return isCallExpression(node) && node.callee && (node as any).callee?.name === 'defineComponent'
}
export function findComponents(ast: AST) {
  const componentsAst: CallExpression[] = []
  traverse(ast, {
    CallExpression({ node }) {
      if (checkIsDefineComponent(node))
        componentsAst.push(node)
    },
  })
  return componentsAst
}
