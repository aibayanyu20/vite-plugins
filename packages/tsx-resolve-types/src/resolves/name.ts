import type { AST } from '../interface'
import { traverse } from '../utils/traverse'

export function resolveName(ast: AST) {
  //  TODO: 从 ast 中解析出 name
  traverse(ast, {

  })
}
