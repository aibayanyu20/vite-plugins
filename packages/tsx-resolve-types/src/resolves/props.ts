import type {
  ArrowFunctionExpression,
  CallExpression,
  Expression,
  FunctionExpression,
  Identifier,
  ObjectExpression,
  Pattern,
  RestElement,
} from '@babel/types'
import {
  identifier,
  isArrowFunctionExpression,
  isAssignmentPattern,
  isFunctionExpression,
  isIdentifier,
  isObjectExpression,
  isObjectProperty,
  objectExpression,
  objectProperty,
} from '@babel/types'
import { extractRuntimeProps } from '@vue/compiler-sfc'
import { parseExpression } from '@babel/parser'
import type { CreateContextType } from '../utils/context'

function getTypeAnnotation(node: Identifier) {
  if (node.typeAnnotation && 'typeAnnotation' in node.typeAnnotation)
    return node.typeAnnotation
}

function addTypes(prop: RestElement | Identifier | Pattern | undefined, ctx: CreateContextType) {
  if (prop && isAssignmentPattern(prop)) {
    if (isIdentifier(prop.left) && !ctx.ctx.propsTypeDecl) {
      const type = getTypeAnnotation(prop.left)
      if (type)
        ctx.ctx.propsTypeDecl = type
    }
    if (prop.right)
      ctx.ctx.propsRuntimeDefaults = prop.right

    return prop.left
  }
  else if (prop && isIdentifier(prop)) {
    const type = getTypeAnnotation(prop)
    if (type)
      ctx.ctx.propsTypeDecl = type
  }
}
function getObjectSetup(expression: ObjectExpression, ctx: CreateContextType) {
  const properties = expression.properties
  const setup = properties.find((property) => {
    return property.type === 'ObjectMethod' && 'name' in property.key && property.key.name === 'setup'
  })
  if (setup?.type === 'ObjectMethod') {
    const params = setup.params
    const prop = params[0]
    const leftAst = addTypes(prop, ctx)
    if (leftAst)
      params[0] = leftAst as any
  }
}

function getPropsTypeToDefine(exp: CallExpression, ctx: CreateContextType) {
  /**
   * 直接在函数的泛型数据中获取
   */
  if (exp.typeParameters) {
    const params = exp.typeParameters?.params
    if (params) {
      const prop = params[0]
      if (prop)
        ctx.ctx.propsTypeDecl = prop
    }
  }
}

function getFuncType(exp: ArrowFunctionExpression | FunctionExpression, ctx: CreateContextType) {
  const params = exp.params
  const prop = params[0]
  const left = addTypes(prop, ctx)
  if (left)
    params[0] = left as any
}

function getPropsStr(ctx: CreateContextType) {
  if (ctx.ctx.propsTypeDecl) {
    // 开始执行函数获取里面的数据信息
    const propStr = extractRuntimeProps(ctx.ctx)
    ctx.ctx.propsTypeDecl = undefined
    ctx.ctx.propsRuntimeDefaults = undefined
    if (propStr) {
      /**
       * 判断生成的代码中是否存在
       */
      if (!ctx.importMergeDefaults)
        (propStr.includes('/*#__PURE__*/_mergeDefaults') || propStr.includes('/*@__PURE__*/_mergeDefaults') || propStr.includes('_mergeDefaults')) && (ctx.importMergeDefaults = true)
      return parseExpression(propStr)
    }
  }
}

function addPropsToFunc(exp: CallExpression, ast: Expression) {
  const arg1 = exp.arguments[1]
  if (arg1) {
    if (isObjectExpression(arg1)) {
      const propsProperty = arg1.properties.find(p => isObjectProperty(p) && 'name' in p.key && p.key.name === 'props')
      if (!propsProperty)
        arg1.properties.unshift(objectProperty(identifier('props'), ast))
    }
  }
  else {
    const obj = objectExpression([objectProperty(identifier('props'), ast)])
    exp.arguments.push(obj)
  }
}
export function resolveProps(expression: CallExpression, ctx: CreateContextType) {
  // 判断的第一个参数的类型
  const props = expression.arguments[0]
  if (!props)
    return
  getPropsTypeToDefine(expression, ctx)
  // 判断第一个参数的类型，是不是一个函数
  if (isFunctionExpression(props) || isArrowFunctionExpression(props)) {
    getFuncType(props, ctx)
    const propAst = getPropsStr(ctx)
    if (propAst)
      addPropsToFunc(expression, propAst)
  }
  else if (props.type === 'ObjectExpression') {
    // 当前是一个对象，就需要判断是否存在一个属性是props
    const propsProperty = props.properties.find(p => isObjectProperty(p) && 'name' in p.key && p.key.name === 'props')
    if (!propsProperty) {
      getObjectSetup(props, ctx)
      const propAst = getPropsStr(ctx)
      if (propAst)
        props.properties.unshift(objectProperty(identifier('props'), propAst))
    }
  }
}
