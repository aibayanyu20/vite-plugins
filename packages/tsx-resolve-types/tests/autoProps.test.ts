import path from 'node:path'
import { describe, expect, it } from 'vitest'
import generate from '@babel/generator'
import ts from 'typescript'
import { registerTS } from '@vue/compiler-sfc'
import { createContext } from '../src/utils/context'
import { findComponents } from '../src/parser'
import { resolveProps } from '../src/resolves/props'
import { checkMergeDefaults } from '../src/utils/checkMergeDefaults'
import { basePath } from './fixtures/constant'
import complexRaw from './fixtures/autoProps/complex.tsx?raw'
import defaultsRaw from './fixtures/autoProps/defaults.tsx?raw'
import exportRaw from './fixtures/autoProps/export.tsx?raw'
import genericRaw from './fixtures/autoProps/generic.tsx?raw'
import inlineRaw from './fixtures/autoProps/inline.tsx?raw'
import libRaw from './fixtures/autoProps/lib.tsx?raw'
import nonTyperRaw from './fixtures/autoProps/nonType.tsx?raw'
import renderRaw from './fixtures/autoProps/render.tsx?raw'
import singleRaw from './fixtures/autoProps/single.tsx?raw'

registerTS(() => ts)

describe('autoProps', () => {
  it('complex', () => {
    const code = compiler(complexRaw, 'complex.tsx')
    expect(code).toMatchSnapshot()
  })

  it('defaults', () => {
    const code = compiler(defaultsRaw, 'defaults.tsx')
    expect(code).toMatchSnapshot()
  })

  it('export', () => {
    const code = compiler(exportRaw, 'export.tsx')
    expect(code).toMatchSnapshot()
  })

  // it('extra', () => {
  //   const code = compiler(extraRaw, 'extra.tsx')
  //   expect(code).toMatchSnapshot()
  // })

  it('generic', () => {
    const code = compiler(genericRaw, 'generic.tsx')
    expect(code).toMatchSnapshot()
  })

  it('inline', () => {
    const code = compiler(inlineRaw, 'inline.tsx')
    expect(code).toMatchSnapshot()
  })

  it('lib', () => {
    const code = compiler(libRaw, 'lib.tsx')
    expect(code).toMatchSnapshot()
  })

  it('nonType', () => {
    const code = compiler(nonTyperRaw, 'nonType.tsx')
    expect(code).toMatchSnapshot()
  })

  it('render', () => {
    const code = compiler(renderRaw, 'render.tsx')
    expect(code).toMatchSnapshot()
  })

  it('single', () => {
    const code = compiler(singleRaw, 'single.tsx')
    expect(code).toMatchSnapshot()
  })
})

function compiler(raw: string, id: string) {
  const ctx = createContext(raw, path.resolve(basePath, `autoProps/${id}`))
  const expression = findComponents(ctx.ast)
  for (const callExpression of expression)
    resolveProps(callExpression, ctx)
  checkMergeDefaults(ctx)

  return generate(ctx.ast).code
}
