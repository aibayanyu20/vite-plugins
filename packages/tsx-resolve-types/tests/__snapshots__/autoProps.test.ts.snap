// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`autoProps > complex 1`] = `
"import { defineComponent } from 'vue';
export interface CommonProps {
  b: string;
}
export interface ComplexProps {
  a: string;
}
export const Complex = defineComponent<CommonProps & ComplexProps>(() => {
  return () => {
    return <div></div>;
  };
}, {
  props: {
    b: {
      type: String,
      required: true
    },
    a: {
      type: String,
      required: true
    }
  }
});
export const Complex1 = defineComponent((_props: CommonProps & ComplexProps) => {
  return () => {
    return <div></div>;
  };
}, {
  props: {
    b: {
      type: String,
      required: true
    },
    a: {
      type: String,
      required: true
    }
  }
});
export const Complex2 = defineComponent((_props: CommonProps & {
  c: string;
}) => {
  return () => {
    return <div></div>;
  };
}, {
  props: {
    b: {
      type: String,
      required: true
    },
    c: {
      type: String,
      required: true
    }
  }
});
export const Complex3 = defineComponent({
  props: {
    b: {
      type: String,
      required: true
    },
    a: {
      type: String,
      required: true
    }
  },
  setup(_props: CommonProps & ComplexProps) {
    return () => {
      return <div></div>;
    };
  }
});
export const Complex4 = defineComponent({
  props: ['a', 'b'],
  setup(_props: CommonProps & ComplexProps) {
    return () => {
      return <div></div>;
    };
  }
});"
`;

exports[`autoProps > defaults 1`] = `
"import { defineComponent } from 'vue';
export const DefaultTsx = defineComponent<{
  a?: string;
}>({
  props: {
    a: {
      type: String,
      required: false,
      default: '1'
    }
  },
  setup(props = {
    a: '1'
  }) {
    return () => {
      return <div></div>;
    };
  }
});"
`;

exports[`autoProps > export 1`] = `
"import { defineComponent } from 'vue';
export interface CommonProps {
  b: string;
}
export default defineComponent<CommonProps>({
  props: {
    b: {
      type: String,
      required: true
    }
  },
  setup() {
    return () => {
      return <div></div>;
    };
  }
});"
`;

exports[`autoProps > generic 1`] = `
"import { defineComponent, ref } from 'vue';
import type { Props } from './typing';
export const Generic = defineComponent(<T extends string | number,>(props: {
  msg: T;
  list: T[];
}) => {
  const count = ref(0);
  return () => <div>
          {props.msg}
          {' '}
          {count.value}
        </div>;
}, {
  props: {
    msg: {
      type: null,
      required: true
    },
    list: {
      type: Array,
      required: true
    }
  }
});
export const Generic1 = defineComponent(<T extends Props | number,>(props: {
  msg: T;
  list: T[];
}) => {
  const count = ref(0);
  return () => <div>
          {props.msg}
          {' '}
          {count.value}
        </div>;
}, {
  props: {
    msg: {
      type: null,
      required: true
    },
    list: {
      type: Array,
      required: true
    }
  }
});"
`;

exports[`autoProps > inline 1`] = `
"import { defineComponent } from 'vue';
export const Inline = defineComponent((_props: {
  c: string;
  a: number;
}) => {
  return () => {
    return <div></div>;
  };
}, {
  props: {
    c: {
      type: String,
      required: true
    },
    a: {
      type: Number,
      required: true
    }
  }
});
export const Inline1 = defineComponent<{
  c: string;
  a: number;
}>(() => {
  return () => {
    return <div></div>;
  };
}, {
  props: {
    c: {
      type: String,
      required: true
    },
    a: {
      type: Number,
      required: true
    }
  }
});
export const Inline2 = defineComponent({
  props: {
    c: {
      type: String,
      required: true
    },
    a: {
      type: Number,
      required: true
    }
  },
  setup(_props: {
    c: string;
    a: number;
  }) {
    return () => {
      return <div></div>;
    };
  }
});
export default defineComponent({
  props: {
    c: {
      type: String,
      required: true
    },
    a: {
      type: Number,
      required: true
    },
    b: {
      type: String,
      required: true
    }
  },
  setup(_props: {
    c: string;
    a: number;
    b: string;
  }) {
    return () => {
      return <div></div>;
    };
  }
});"
`;

exports[`autoProps > lib 1`] = `
"import type { TableProps } from 'ant-design-vue';
import { defineComponent } from 'vue';
export type MyTableProps = TableProps & {
  sss?: string;
};
export const LibTest = defineComponent<MyTableProps>(() => {
  return () => {
    return <div></div>;
  };
}, {
  props: {
    dropdownPrefixCls: {
      type: String,
      required: false
    },
    dataSource: {
      type: Array,
      required: false
    },
    columns: {
      type: null,
      required: false
    },
    pagination: {
      type: [Boolean, Object],
      required: false
    },
    loading: {
      type: [Boolean, Object],
      required: false
    },
    size: {
      type: null,
      required: false
    },
    bordered: {
      type: Boolean,
      required: false
    },
    locale: {
      type: Object,
      required: false
    },
    onChange: {
      type: Function,
      required: false
    },
    onResizeColumn: {
      type: Function,
      required: false
    },
    rowSelection: {
      type: Object,
      required: false
    },
    getPopupContainer: {
      type: Function,
      required: false
    },
    scroll: {
      type: Object,
      required: false
    },
    sortDirections: {
      type: Array,
      required: false
    },
    showSorterTooltip: {
      type: [Boolean, Object],
      required: false
    },
    prefixCls: {
      type: String,
      required: false
    },
    rowKey: {
      type: null,
      required: false
    },
    tableLayout: {
      type: String,
      required: false
    },
    rowClassName: {
      type: null,
      required: false
    },
    title: {
      type: null,
      required: false
    },
    footer: {
      type: null,
      required: false
    },
    id: {
      type: String,
      required: false
    },
    showHeader: {
      type: Boolean,
      required: false
    },
    components: {
      type: Object,
      required: false
    },
    customRow: {
      type: null,
      required: false
    },
    customHeaderRow: {
      type: null,
      required: false
    },
    direction: {
      type: String,
      required: false
    },
    expandFixed: {
      type: [String, Boolean],
      required: false
    },
    expandColumnWidth: {
      type: Number,
      required: false
    },
    expandedRowKeys: {
      type: Array,
      required: false
    },
    defaultExpandedRowKeys: {
      type: Array,
      required: false
    },
    expandedRowRender: {
      type: null,
      required: false
    },
    expandRowByClick: {
      type: Boolean,
      required: false
    },
    expandIcon: {
      type: null,
      required: false
    },
    onExpand: {
      type: Function,
      required: false
    },
    onExpandedRowsChange: {
      type: Function,
      required: false
    },
    defaultExpandAllRows: {
      type: Boolean,
      required: false
    },
    indentSize: {
      type: Number,
      required: false
    },
    expandIconColumnIndex: {
      type: Number,
      required: false
    },
    showExpandColumn: {
      type: Boolean,
      required: false
    },
    expandedRowClassName: {
      type: null,
      required: false
    },
    childrenColumnName: {
      type: String,
      required: false
    },
    rowExpandable: {
      type: Function,
      required: false
    },
    sticky: {
      type: [Boolean, Object],
      required: false
    },
    transformCellText: {
      type: null,
      required: false
    },
    sss: {
      type: String,
      required: false
    }
  }
});"
`;

exports[`autoProps > nonType 1`] = `
"import { defineComponent } from 'vue';
export const NonType = defineComponent(() => {
  return () => {
    return <div></div>;
  };
});"
`;

exports[`autoProps > render 1`] = `
"import { defineComponent } from 'vue';
export default defineComponent({
  setup() {
    return {};
  },
  render() {
    return <div></div>;
  }
});
export const A = defineComponent({
  props: {
    a: {
      type: String,
      required: true
    }
  },
  setup(_props: {
    a: string;
  }) {
    return {};
  },
  render() {
    return <div></div>;
  }
});"
`;

exports[`autoProps > single 1`] = `
"import { defineComponent } from 'vue';
import type { Props } from './typing';
export interface Props1 {
  foo1: string;
  bar1: number;
}
export type Props2 = Props & Props1;
export type Props3 = Pick<Props2, 'foo1' | 'foo'>;
export const Single = defineComponent<Props>(() => {
  return () => {
    return <div></div>;
  };
}, {
  props: {
    foo: {
      type: String,
      required: true
    },
    bar: {
      type: Number,
      required: true
    }
  }
});
export const Single2 = defineComponent((_props: Props2) => {
  return () => {
    return <div></div>;
  };
}, {
  props: {
    foo: {
      type: String,
      required: true
    },
    bar: {
      type: Number,
      required: true
    },
    foo1: {
      type: String,
      required: true
    },
    bar1: {
      type: Number,
      required: true
    }
  }
});
export const Single3 = defineComponent({
  props: {
    foo1: {
      type: String,
      required: true
    },
    foo: {
      type: String,
      required: true
    }
  },
  setup(_props: Props3) {
    return () => {
      return <div></div>;
    };
  }
});"
`;
